lucet-runtime/lucet-runtime-internals/src/region/mmap.rs

```diff
@@ -80,38 +80,42 @@ impl RegionInternal for MmapRegion {
         module: Arc<dyn Module>,
         embed_ctx: CtxMap,
         heap_memory_size_limit: usize,
+        alloc_strategy: AllocStrategy,
     ) -> Result<InstanceHandle, Error> {
-        let custom_limits;
-        let mut limits = self.get_limits();
-
-        // Affirm that the module, if instantiated, would not violate
-        // any runtime memory limits.
-        match heap_memory_size_limit.cmp(&limits.heap_memory_size) {
-            Ordering::Less => {
-                // The supplied heap_memory_size is smaller than
-                // default. Augment the limits with this custom value
-                // so that it may be validated.
-                custom_limits = Limits {
-                    heap_memory_size: heap_memory_size_limit,
-                    ..*limits
-                };
-                limits = &custom_limits;
+        let limits = self.get_limits();
+        module.validate_runtime_spec(&limits, heap_memory_size_limit)?;
+
+        let slot;
+
+        match alloc_strategy {
```

Comment:
Could you abstract this into a method on `AllocStrategy`, so that the logic could be reused with `UffdRegion`? It seems like you could get away with something like:

```rust
impl AllocStrategy {
    fn next(&mut self, free_slots: usize) -> usize { ... }
}
```

Then the region could handle its own `swap_remove` logic, etc.

Reply:
Neat.  I'll give this a try.

Reply:
One small worry.  What if `free_slots` is 0?  I'm tempted to return a `Result` instead of a usize.  What are your thoughts?

Reply:
Good idea!